modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/service/impl/ModelServiceImpl.java

```diff
@@ -42,26 +44,55 @@
     @Autowired
     private ModelRepository modelRepository;
 
+    @Autowired
+    private MetricRepository metricRepository;
+
     @Override
     public List<ModelDto> getModels(String metricKey) {
         return modelRepositoryCustom.findModels(metricKey);
     }
 
     @Override
-    public void addModelParams(ModelParams modelParams) {
+    public void addHyperparams(String uuid, Hyperparams hyperparams) {
 
-        Model model = new Model(modelParams.getModelUUID(), modelParams.getHyperparams());
-        Metric metric = new Metric(modelParams.getMetricKey());
-        model.getMetrics().add(metric);
-        metric.getModels().add(model);
-        modelRepository.save(model);
+        AssertUtil.notNull(hyperparams.getHyperparams(), "Hyperparams can't be null");
+        AssertUtil.notNull(hyperparams.getMetricKey(), "Metric key can't be null");
+        AssertUtil.notNull(uuid, "Model UUID can't be null");
+
+        Integer metricId = metricRepository.findIdByMetricKey(hyperparams.getMetricKey());
+        Integer modelId = modelRepository.findIdByModelUUID(uuid);
+
+        Metric metric;
+        if (metricId == null) {
+            metric = new Metric(hyperparams.getMetricKey());
+        } else {
+            metric = metricRepository.getOne(metricId);
+        }
+
+        Model model;
+        if (modelId == null) {
+            model = new Model(uuid, hyperparams.getHyperparams());
+        } else {
+            model = modelRepository.getOne(modelId);
+        }
+
+        if (metricId == null || modelId == null) {
+            model.getMetrics().add(metric);
+            metric.getModels().add(model);
+            modelRepository.save(model);
+        } else {
+            throw new RuntimeException("Metric-Model mapping already exits");
+        }
     }
 
     @Override
-    public void markToRebuild(String modelUUID, String metricKey, Boolean toRebuild) {
+    public void markToRebuild(String modelUUID, String metricKey, boolean toRebuild) {
```

Comment:
Let's chat about this one in person. I am not convinced that we want some external agent to mark these models for rebuild, at least at this time. Presumably the rebuild triggers would either be cron-based or else performance-based.

For a cron-based approach I think we'd want to trigger off of a declared schedule, and we may as well put that schedule in the model service instead of assuming some external thing know the schedule and call `ModelService.markToRebuild()`. (No reason to push the problem back.)

For the performance-based approach, we would need some good way to measure model performance, and we don't yet have that. (I don't any longer think RMSE is the right thing here--that's prediction model performance, not classification model performance, and it's trivial to make the predictions _perfect_ and have bad classifications.)

So my suggestion is to get rid of `markToRebuild` for now, and let's figure out how to define a rebuild schedule.